Offline Node Support

Daniel-VDM opened the pull request:
This feature is needed for a correct deployment of a rosetta 'offline' node. That said, it is a perfectly reasonable feature to have on its own - rosetta notwithstanding. I imagine that it would great for debugging / devs. 

One can run a node in offline mode by adding the `--run.offline` option. This option has been tested to be backwards compatible with older config files. It defaults to an 'online' setting if the option/config is not provided. 

An offline node simply means that it does not sync & is not subscribed to any p2p topic or has any p2p peers. 
I've had to keep some service instantiations to ensure other services don't crash from a nil pointer.

JackyWYX commented:
I would suggest add the following logic:

1. If running in offline mode, force p2p to listen to address `127.0.0.1 (p2p/host.go:68)` (It might be better to add a field `IP` in `p2pConfig` `cmd/harmony/config.go:45`). 
2. If run in offline mode, instruct `node.SyncingPeerProvider` to use local network.

This can further prevent node from using network. Else the node is still using libp2p for peer discovery. Do you think this is reasonable?

LeoHChen reviewed (commented):
reasonable request by @JackyWYX , please fix them. thanks.

Daniel-VDM commented:
> I would suggest add the following logic:
> 
> 1. If running in offline mode, force p2p to listen to address `127.0.0.1 (p2p/host.go:68)` (It might be better to add a field `IP` in `p2pConfig` `cmd/harmony/config.go:45`).
> 2. If run in offline mode, instruct `node.SyncingPeerProvider` to use local network.
> 
> This can further prevent node from using network. Else the node is still using libp2p for peer discovery. Do you think this is reasonable?

Sure i

> 2\. SyncingPeerProvider

Sure seems fair enough, added this logic and fixed related tests along the way.

JackyWYX commented on internal/configs/node/network.go:
```diff
@@ -39,6 +39,10 @@ const (
 )
 
 const (
+	// DefaultLocalListenIP ..
```
I would suggest to finish these lines. Using ... is not good coding practice (I know the function name is already very elaborative, but would suggest to explain it in human language).

Daniel-VDM replied:
Ill remove it then

Daniel-VDM replied:
Hmm.. sure ill add some comments

JackyWYX replied:
Sure. I definitely agree with you. A good code shall be self explanatory without any comments. But this is syntax required here :(

JackyWYX commented on node/node.go:
```diff
@@ -599,6 +601,10 @@ func (node *Node) Start() error {
 	isThisNodeAnExplorerNode := node.NodeConfig.Role() == nodeconfig.ExplorerNode
 
+	if node.NodeConfig.IsOffline {
+		select {} // sleep until program is terminated
```
Instead of adding more logic in this already huge function, we can simply leave this function unchanged since we have already changed p2p ip to local address. Otherwise, I would suggest you to write a `runOfflineNode` which is simply `select {}` in the if condition before enter `node.Start()` since `node.Start` is all about p2p pubsub message handling.

Daniel-VDM replied:
I'll do the ladder approach then..

JackyWYX commented on cmd/harmony/main.go:
```diff
@@ -178,6 +178,10 @@ func getHarmonyConfig(cmd *cobra.Command) (harmonyConfig, error) {
 		return harmonyConfig{}, err
 	}
 
+	if config.General.IsOffline {
```
From my point of view, this piece of code are not supposed to be here. Can we move this piece of code to `applyGeneralFlags`, and then in `validateHarmonyConfig`, return an error when the p2p IP and offline settings is not consistent?

JackyWYX commented:
I think this is a public variable right? Comments are required for public
stuff😂

On Fri, Oct 9, 2020 at 2:09 AM Daniel Van Der Maden <
notifications@github.com> wrote:

> *@Daniel-VDM* commented on this pull request.
> ------------------------------
>
> In internal/configs/node/network.go
> <https://github.com/harmony-one/harmony/pull/3383#discussion_r502293588>:
>
> > @@ -39,6 +39,10 @@ const (
>  )
>
>  const (
> +	// DefaultLocalListenIP ..
>
> Ill remove it then
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/harmony-one/harmony/pull/3383#discussion_r502293588>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJSNLCEZ3T6HV2C55PMA2VTSJ3HLNANCNFSM4SHISFHQ>
> .
>

Daniel-VDM commented:
> I think this is a public variable right? Comments are required for public stuff😂
> […](#)
> On Fri, Oct 9, 2020 at 2:09 AM Daniel Van Der Maden < ***@***.***> wrote: *@Daniel-VDM* commented on this pull request. ------------------------------ In internal/configs/node/network.go <[#3383 (comment)](https://github.com/harmony-one/harmony/pull/3383#discussion_r502293588)>: > @@ -39,6 +39,10 @@ const ( ) const ( + // DefaultLocalListenIP .. Ill remove it then — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub <[#3383 (comment)](https://github.com/harmony-one/harmony/pull/3383#discussion_r502293588)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AJSNLCEZ3T6HV2C55PMA2VTSJ3HLNANCNFSM4SHISFHQ> .

Yea... updated the comment right after

JackyWYX commented on cmd/harmony/flags.go:
```diff
@@ -271,6 +278,10 @@ func applyGeneralFlags(cmd *cobra.Command, config *harmonyConfig) {
 	}
+
+	if cli.IsFlagChanged(cmd, isOfflineFlag) {
+		config.General.IsOffline = cli.GetBoolFlagValue(cmd, isOfflineFlag)
```
Shall we add here?

```
isOffline : =cli.GetBoolFlagValue(cmd, isOfflineFlag)
config.General.IsOffline = isOffline
if isOffline {
     config.P2P.IP = nodeconfig. DefaultLocalListenIP
}
```

Daniel-VDM replied:
Hmm, well the current impl already accounts for this. 
If we want to implicitly set the IP, we should prob do it in `applyP2PFlags`.

Daniel-VDM replied:
Prob makes sense to implicitly set the IP, ill update.

JackyWYX reviewed (approved):
Nice!
